optimum/fx/parallelization/passes.py

```diff
@@ -0,0 +1,449 @@
+                static config to include instructions which persists the whole process.
+
+        Returns:
+            GraphModule: graph module after processed by the current pass.
+        """
+        raise NotImplementedError
+
+    def __call__(self, graph_module: GraphModule, ctx: ParallelExecutionCtx, config: Config) -> GraphModule:
+        graph_module = self.run(graph_module, ctx=ctx, config=config)
+        if config.lint_and_recompile:
+            graph_module.graph.lint()
+            graph_module.recompile()
+        return graph_module
+
+
+class AnalyzeBase(PassBase):
+    """
+    Base class for passes which only run for analytical purposes and preserve graph structure
+    during processing. Analytical passes are often prerequisite passes which provide information
+    for passes later on to actually change the graph.
+
+    Passes inheriting from `AnalyBase` places the class signature as a meta key in `node.meta`,
+    which is a dict storing meta information related with a fx Node, such as the shape and dtype of
+    output. Look-up APIs are exposed as classmethod so that passes using them won't need to create
+    concrete instances.
+    """
+    @classmethod
+    def meta_key(cls) -> str:
+        # place class-wise unique meta_key in `meta` to prevent duplicate fields
+        return cls.signature()
+    
+    @classmethod
+    def get_stored_field_info(cls, node: Node, field: Any, must_have: bool = False) -> Any:
+        if not cls.already_executed_per_node(node):
+            if not must_have:
+                return None
+            else:
+                raise RuntimeError(
+                    f"Can't find information related with {cls.__name__} in the current node `{node}` "
+                    f"make sure {cls.__name__} has run and marked it"
+                )
+        
+        info : Dict[Any, Any] = node.meta[cls.meta_key()]
+        if field not in info:
+            raise KeyError(f"Invalid query field {field} for {cls.__name__}, valid fields are {list(info.keys())}")
+
+        return info[field]
+    
+    @classmethod
+    def already_executed_per_node(cls, node: Node) -> None:
+        return cls.meta_key() in node.meta
+    
+    def place_marker_per_node(self, node: Node, info: Dict[Any, Any]) -> None:
+        if self.already_executed_per_node(node):
+            raise RuntimeError(
+                f"Node {node} has already been marked by the current pass, check if "
+                "the current pass has already been executed in the pipeline"
+            )
+
+        node.meta[self.meta_key()] = info
+
+    def clear_marker_per_node(self, node: Node) -> None:
+        key = self.meta_key()
+        if key in node.meta:
+            node.meta.pop(key)
+
+    def clean_all(self, graph_module: GraphModule) -> None:
+        g : Graph = graph_module.graph
+        for node in g.nodes:
+            self.clear_marker_per_node(node)
+
+
+class ParallelLinearAnnotatePass(AnalyzeBase):
```

Comment:
I don't get the name parallel here? Isn't it more like successive?

Reply:
it actually means annotate some layers to be their parallel counterparts